[Infra] WebClient 설정 환경 구성

vwx12345 opened the pull request:
### 💻 작업 내용

- `spring-boot-starter-webflux` 의존성 추가
- `WebClient.Builder` 빈 등록을 위한 `WebClientConfig` 클래스 생성
- 요청/응답 로깅용 ExchangeFilterFunction 설정 추가

### ✨ 리뷰 포인트


### 📝 메모
- 해당 설정은 내부 API 호출 시 WebClient를 사용하기 위한 기반 작업입니다.
- 현재는 동기식 .block() 사용을 가정하고 있으나, 향후 비동기 처리 확장을 고려해 WebClient로 통일했습니다.
- webClinet는 기본적으로 외부 API용이라 현재 에러코드를 외부로 정의했습니다.
- 
### 🎯 관련 이슈
closed #64

vwx12345 committed:
feat(infra): WebClient 사용을 위한 webflux 의존성 추가

vwx12345 committed:
feat(infra): WebClient.Builder 빈 등록을 위한 설정 클래스 추가

codingbaraGo commented:
혹시 exception 핸들링에 관한 아이디어도 있으신가요?
webflux를 사용하면 response 형식이 달라져야해서 exception 핸들링도 따로 처리해주어야 할듯 합니다.
webflux 도입 pr에 exception 핸들러까지 넣어주는게 좋을것 같은데 어떻게 생각하시나요?

vwx12345 commented:
> 혹시 exception 핸들링에 관한 아이디어도 있으신가요? webflux를 사용하면 response 형식이 달라져야해서 exception 핸들링도 따로 처리해주어야 할듯 합니다. webflux 도입 pr에 exception 핸들러까지 넣어주는게 좋을것 같은데 어떻게 생각하시나요?

확인했습니다. 그 부분도 추가해서 다시 pr하겠습니다.

vwx12345 committed:
feat(infra): webClient 기본 로깅 추가

- WebClient에 요청/응답 헤더 로그 필터 추가
- 요청 시 HTTP 메서드, URL, 헤더 정보 출력
- 응답 시 응답 헤더 출력
- body는 출력하지 않도록 기본 설정 유지

vwx12345 committed:
feat(infra): WebClientErrorHandler 구현

- WebClient의 4xx, 5xx 응답에 대한 공통 예외 처리 유틸 클래스 추가
- 응답 본문과 상태 코드를 기반으로 ClientErrorException, ServerErrorException 발생

vwx12345 committed:
feat(global): WebClient 예외 대응 GlobalExceptionHandler 및 ErrorCode 정의

- WebClient에서 발생하는 ClientErrorException, ServerErrorException 핸들러 추가
- ErrorCode에 EXTERNAL_CLIENT_ERROR, EXTERNAL_SERVER_ERROR 항목 정의

codingbaraGo commented on src/main/java/com/simpleboard/board/global/exception/webClient/WebClientErrorHandler.java:
```diff
@@ -0,0 +1,47 @@
+ * <p>ClientResponse로부터 상태 코드와 응답 본문을 추출하여,
+ * 적절한 예외 로그를 기록하고 RuntimeException 하위 커스텀 예외를 발생시킵니다.
+ */
+public class WebClientErrorHandler {
+
+  public static Mono<? extends Throwable> handle4xx(ClientResponse response) {
+    return response.bodyToMono(String.class)
+        .defaultIfEmpty("")
```
제가 webflux를 잘 몰라서 질문드립니다.
지금 이 클래스가 에러 핸들러라는것은 알겠는데, 어떻게 에러 핸들러로서 작동하는건가요?
config 클래스에 따로 등록해주는것도 아닌듯 하고, 특정 인터페이스를 implement 받는것도 아닌데 그냥 이렇게 정의만 해 놓아도 error handler로서 작동을 하나요?
이름이 트리거인가요?

vwx12345 replied:
> 제가 webflux를 잘 몰라서 질문드립니다. 지금 이 클래스가 에러 핸들러라는것은 알겠는데, 어떻게 에러 핸들러로서 작동하는건가요? config 클래스에 따로 등록해주는것도 아닌듯 하고, 특정 인터페이스를 implement 받는것도 아닌데 그냥 이렇게 정의만 해 놓아도 error handler로서 작동을 하나요? 이름이 트리거인가요?

Spring WebFlux에는 @ControllerAdvice나 전역 예외 핸들러처럼 WebClient 호출에 대한 전역 오류 필터가 기본적으로는 없습니다. 따라서  WebClient는 HTTP 응답 코드별로 수동으로 예외를 처리하도록 API를 설계해두었기 때문에, 아래처럼 .onStatus(...)로 명시적으로 연결해야 합니다.

webClient.get()
    .uri("/api")
    .retrieve()
    .onStatus(HttpStatus::is4xxClientError, WebClientErrorHandler::handle4xx)
    .onStatus(HttpStatus::is5xxServerError, WebClientErrorHandler::handle5xx)
    .bodyToMono(SomeResponse.class)
    .block();

codingbaraGo replied:
이 핸들러는 WebClient에 대한 에러 핸들러인가요? 아니면 webflux 컨트롤러에 대한 에러 핸들러인가요?
저는 WebFlux 컨트롤러를 도입한다 생각해서 Webflux 컨트롤러에 대한 에러 핸들러라 생각했는데 webclient만 도입한다 하셔서요.

vwx12345 replied:
> 이 핸들러는 WebClient에 대한 에러 핸들러인가요? 아니면 webflux 컨트롤러에 대한 에러 핸들러인가요? 저는 WebFlux 컨트롤러를 도입한다 생각해서 Webflux 컨트롤러에 대한 에러 핸들러라 생각했는데 webclient만 도입한다 하셔서요.

컨트롤러는 기존의 rest controller를 사용합니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/global/exception/webClient/WebClientErrorHandler.java:
```diff
@@ -0,0 +1,47 @@
+
+          return Mono.error(new ClientErrorException(status, body));
+        });
+  }
```
여기에서는 커스텀 예외 클래스인 ServiceException을 던지고 있는데, 여기서 return된 exception들은 어디로 가나요?
근데 exception인데 throw가 아니라 return으로 전달하는군요.
이런 부분은 webflux 만의 특징인가요?

vwx12345 replied:
> 여기에서는 커스텀 예외 클래스인 ServiceException을 던지고 있는데, 여기서 return된 exception들은 어디로 가나요? 근데 exception인데 throw가 아니라 return으로 전달하는군요. 이런 부분은 webflux 만의 특징인가요?

해당 부분은 WebFlux의 특징이 맞습니다.
Mono는 비동기 스트림의 흐름을 표현하는 Publisher이기 때문에, 예외를 throw하는 것이 아니라 Mono.error()를 통해 에러를 포함한 스트림으로 반환합니다.
이렇게 반환된 에러는 실제로 해당 Mono가 subscribe()되거나, 저희처럼 .block()을 호출해 동기적으로 처리하는 시점에 정의한 커스텀 예외가 발생하게 됩니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/global/exception/webClient/ClientErrorException.java:
```diff
@@ -0,0 +1,19 @@
+ * <p>WebClient 4xx 응답 처리용 커스텀 예외
+ */
+@Getter
+public class ClientErrorException extends RuntimeException {
```
클래스 이름에 내부 api 관련 exception, 혹은 webflux 관련 exception이라는 것을 명시하는게 어떤가요?
그리고 지금 주석만 봤을 때는 해당 exception이 어떤 상황에서 발생하는지 잘 모르겠습니다.
spring webflux로 구축된 controller 하위에서 발생한 모든 4xx exception을 대리하는 클래스인가요?

vwx12345 replied:
> 클래스 이름에 내부 api 관련 exception, 혹은 webflux 관련 exception이라는 것을 명시하는게 어떤가요? 그리고 지금 주석만 봤을 때는 해당 exception이 어떤 상황에서 발생하는지 잘 모르겠습니다. spring webflux로 구축된 controller 하위에서 발생한 모든 4xx exception을 대리하는 클래스인가요?

해당 예외는 Spring WebFlux 기반의 컨트롤러에서 발생한 예외가 아닌,
WebClient를 통해 외부 또는 내부 API에 HTTP 요청을 보냈을 때, 응답으로 오는 4xx 상태 코드에 대해 처리하는 용도입니다.

codingbaraGo replied:
넵 알겠습니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/global/exception/webClient/ServerErrorException.java:
```diff
@@ -0,0 +1,19 @@
+ * <p>WebClient 5xx 응답 처리용 커스텀 예외
+ */
+@Getter
+public class ServerErrorException extends RuntimeException {
```
ClientErrorException 클래스에 단 코멘트와 같은점이 궁금하네요.

vwx12345 replied:
> ClientErrorException 클래스에 단 코멘트와 같은점이 궁금하네요.

이 코드는 WebClient를 통해 외부 또는 내부 API에 HTTP 요청을 보냈을 때, 응답으로 오는 5xx 상태 코드에 대해 처리하는 용도입니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/global/exception/GlobalExceptionHandler.java:
```diff
@@ -90,7 +92,21 @@ public ResponseEntity<ErrorResponse> handleNoHandler(
   }
 
-  /** 7. 예상치 못한 모든 예외 – 500 Fallback */
+  /** 7. WebClient – 4xx 예외 처리 */
+  @ExceptionHandler(ClientErrorException.class)
+  public ResponseEntity<ErrorResponse> handleClientError(ClientErrorException ex, HttpServletRequest req) {
+    logException(req, ex);
+    return buildResponse(ErrorCode.EXTERNAL_CLIENT_ERROR);
+  }
+
+  /** 8. WebClient – 5xx 예외 처리 */
+  @ExceptionHandler(ServerErrorException.class)
+  public ResponseEntity<ErrorResponse> handleServerError(ServerErrorException ex, HttpServletRequest req) {
+    logException(req, ex);
+    return buildResponse(ErrorCode.EXTERNAL_SERVER_ERROR);
+  }
```
ClientErrorException과 ServerErrorException 클래스는 둘다 ServiceException 클래스를 상속받은 예외들이라 그냥 놔둬도 handleCustomException 메서드에서 처리해줄텐데 별개의 메서드로 분리한 이유가 있나요?
로직은 같은듯 한데 메서드로 빼신 이유가 궁금하네요.

vwx12345 replied:
> ClientErrorException과 ServerErrorException 클래스는 둘다 ServiceException 클래스를 상속받은 예외들이라 그냥 놔둬도 handleCustomException 메서드에서 처리해줄텐데 별개의 메서드로 분리한 이유가 있나요? 로직은 같은듯 한데 메서드로 빼신 이유가 궁금하네요.

ClientErrorException과 ServerErrorException 클래스는 둘 다 RuntimeException 클래스를 상속받아서 별도의 Handler를 만들어 주었습니다. 이 부분은 해당 예외들이 ServiceException을 상속 받을지, RuntimeException을 상속 받을지 좀 더 생각해볼 부분은 있을 것 같네요.

codingbaraGo replied:
아 제가 잘못봤네요
Service exception이 아니라 runtime exception이었군요

codingbaraGo reviewed (commented):
webflux를 찾아보니 spring mvc와 같은 레벨의 기술이네요.
단순히 controller 구현체중 하나라고 생각했는데 생각보다 큰 기술이라 놀랬습니다.
이거... 안에 들어가는 개념들이 spring mvc만큼이나 많은것같은데 사용해도 괜찮을까요?

webflux는 reactive programming model이라 논블로킹, 비동기로 동작하는듯 한데,
내부 api의 presentation 계층에서 spring mvc 대신 spring webflux를 사용하였을 때 장점이 있을까요?

vwx12345 commented:
> webflux를 찾아보니 spring mvc와 같은 레벨의 기술이네요. 단순히 controller 구현체중 하나라고 생각했는데 생각보다 큰 기술이라 놀랬습니다. 이거... 안에 들어가는 개념들이 spring mvc만큼이나 많은것같은데 사용해도 괜찮을까요?
> 
> webflux는 reactive programming model이라 논블로킹, 비동기로 동작하는듯 한데, 내부 api의 presentation 계층에서 spring mvc 대신 spring webflux를 사용하였을 때 장점이 있을까요?

WebFlux 전체를 사용하는 건 아니고, 내부 API 요청을 위해 WebClient만 도입한 상황입니다.
기존에는 Spring MVC 환경에서 RestTemplate을 사용했었는데,
Spring 5.0부터는 이게 유지보수 상태로 전환됐고, 추후에는 deprecated될 예정이라는 걸 보고 WebClient로 도입하게 됐습니다.

codingbaraGo commented:
> > webflux를 찾아보니 spring mvc와 같은 레벨의 기술이네요. 단순히 controller 구현체중 하나라고 생각했는데 생각보다 큰 기술이라 놀랬습니다. 이거... 안에 들어가는 개념들이 spring mvc만큼이나 많은것같은데 사용해도 괜찮을까요?
> > webflux는 reactive programming model이라 논블로킹, 비동기로 동작하는듯 한데, 내부 api의 presentation 계층에서 spring mvc 대신 spring webflux를 사용하였을 때 장점이 있을까요?
> 
> WebFlux 전체를 사용하는 건 아니고, 내부 API 요청을 위해 WebClient만 도입한 상황입니다. 기존에는 Spring MVC 환경에서 RestTemplate을 사용했었는데, Spring 5.0부터는 이게 유지보수 상태로 전환됐고, 추후에는 deprecated될 예정이라는 걸 보고 WebClient로 도입하게 됐습니다.

아하 알겠습니다.
그렇다면 outbound port에서는 webclient를 사용하고, presentation 계층(inbound port)은 여전히 rest controller만을 사용하는 정책이라 생각하면 될까요?

vwx12345 commented:
> > > webflux를 찾아보니 spring mvc와 같은 레벨의 기술이네요. 단순히 controller 구현체중 하나라고 생각했는데 생각보다 큰 기술이라 놀랬습니다. 이거... 안에 들어가는 개념들이 spring mvc만큼이나 많은것같은데 사용해도 괜찮을까요?
> > > webflux는 reactive programming model이라 논블로킹, 비동기로 동작하는듯 한데, 내부 api의 presentation 계층에서 spring mvc 대신 spring webflux를 사용하였을 때 장점이 있을까요?
> > 
> > 
> > WebFlux 전체를 사용하는 건 아니고, 내부 API 요청을 위해 WebClient만 도입한 상황입니다. 기존에는 Spring MVC 환경에서 RestTemplate을 사용했었는데, Spring 5.0부터는 이게 유지보수 상태로 전환됐고, 추후에는 deprecated될 예정이라는 걸 보고 WebClient로 도입하게 됐습니다.
> 
> 아하 알겠습니다. 그렇다면 outbound port에서는 webclient를 사용하고, presentation 계층(inbound port)은 여전히 rest controller만을 사용하는 정책이라 생각하면 될까요?

네, 맞습니다. 요청을 보내는 것만 webClient를 사용하는 것이고 받는 것은 기존의 rest controller 만을 사용합니다.

vwx12345 committed:
chore(webclient): 패키지 이름 소문자로 변경

codingbaraGo reviewed (approved):
고생하셨습니다.

vwx12345 commented:
> 고생하셨습니다.

그냥 사용하기에는 테스트를 해보지는 못했어서 테스트 코드 작성해서 테스트해보고 다시 pr 올리겠습니다.

vwx12345 committed:
refactor(webclient): 이름 변경 및 기본 메시지 추가

vwx12345 committed:
test(webclient): WebClient 예외 처리 테스트 추가 (4xx/5xx + 바디 유무)

- WebClientErrorHandler 테스트용 서비스 구현
- MockWebServer 의존성 주입
- MockWebServer 기반 테스트 코드 작성

vwx12345 committed:
Merge remote-tracking branch 'origin' into feat/web-flux-setting/#64

codingbaraGo reviewed (approved):
테스트도 깔끔하게 잘 짜셨네요.
고생하셨습니다!
